PSKoans/Koans/Constructs and Patterns/AboutErrorHandling.Koans.ps1

```diff
@@ -0,0 +1,268 @@
+                    'PSKoans.TestWriteErrorMethod',
+                    [System.Management.Automation.ErrorCategory]::LimitsExceeded,
+                    { Secret-ScriptBlock }
+                )
+
+                $PSCmdlet.WriteError($ErrorRecord)
+            }
+        }
+
+        It 'is emitted with the Write-Error cmdlet' {
+            # The redirection operators can be used to more easily retrieve non-standard output, like errors
+            $Error = Test-SimpleWriteError 2>&1
+
+            '__' | Should -Be $Error.GetType().Name
+        }
+
+        It 'can be customized in detail with the Write-Error cmdlet' {
+            $Record = Write-DetailedError 2>&1
+
+            '__' | Should -Be $Record.Exception.Message
+            '__.__' | Should -Be $Record.FullyQualifiedErrorId
+            '__' | Should -Be $Record.TargetObject.Name
+            '__' | Should -Be $Record.ErrorDetails.RecommendedAction
+        }
+
+        It 'is created by WriteError()' {
+            $Record = Write-ErrorWithMethod 2>&1
+            '__' | Should -Be $Record.FullyQualifiedErrorId
+            '__' | Should -Be $Record.TargetObject.ToString()
+        }
+    }
+
+    Context 'Terminating Errors' {
+        <#
+            Terminating errors are generated in a few ways:
+              - The "throw" keyword.
+              - The $PSCmdlet.ThrowTerminatingError() method.
+              - Using -ErrorAction Stop on any cmdlet or advanced function that generates a non-terminating error.
+
+            Note that -ErrorAction only affects errors which are created as non-terminating. Errors that are
+            created as terminating cannot be guided via the -ErrorAction common parameter and will always be
+            regarded as terminating.
+
+            Terminating errors will terminate all further execution at the scope they occur in, and pass control
+            back to the next scope up. The error will be propagated upwards until either it reaches the console
+            and is displayed (at which point all currently-executing scripts will stop) or it triggers a "catch"
+            block which will handle the error.
+
+            The typical try/catch pattern looks like this:
+
+                try {
+                    # Code that might error.
+                }
+                catch [ExceptionType] {
+                    # Code that handles error.
+                    # In this block, $_ refers to the current ErrorRecord.
+                }
+                finally {
+
+                }
+
+            Note that specifying the exception type is optional. A "generic" catch block (no exception type
+            specified) will simply catch everything. If you know the errors you're looking for, it's often
+            best to only catch the ones you're expecting. If something else happens, you might make matters
+            worse by handling that the same way as the errors you were expecting.
```

Comment:
A short example of bad error handling practice might be useful here e.g. `try { throw [InvalidOperationException]"Cannot write into file log.txt" } catch { Write-Host "Cannot connect to network" } `

Reply:
That's a good idea, thanks!